### PR TITLE
[SFSRAP-37] Modified generator to not require the JDK

### PR DESCRIPTION
**Pull request by duncanpMS**
I've added a new builder class that updates an existing jar rather than building a new one.
I haven't removed the old code yet to make reviewing the code more manageable.

TODO:
* remove dead code
* handle long lines in the manifest

I've manually tested generating and installing two new plugins, then running an analysis, with and without an embedded SQALE file. All works as expected so far.

### CODE REVIEW

File: RoslynPluginGenerator/AnalyzerPluginGenerator.cs [hunk 1]
```diff
@@ -31,30 +37,6 @@ public class AnalyzerPluginGenerator
...
-        };
-
-        private const string AnalyzerId_Token = "[ROSLYN_ANALYZER_ID]";
-        private const string RuleNamespace_Token = "[ROSLYN_RULE_NAMESPACE]";
-        private const string PackageId_Token = "[ROSLYN_NUGET_PACKAGE_ID]";
-        private const string PackageVersion_Token = "[ROSLYN_NUGET_PACKAGE_VERSION]";
-        private const string StaticResourceName_Token = "[ROSLYN_STATIC_RESOURCENAME]";
-        private const string PluginKey_Token = "[ROSLYN_PLUGIN_KEY]";
-        private const string PluginVersion_Token = "[ROSLYN_PLUGIN_VERSION]";
-
```
Comment (duncanpMS) [inline RoslynPluginGenerator/AnalyzerPluginGenerator.cs]: This class used to inject values source code - no longer required.

File: RoslynPluginGenerator/AnalyzerPluginGenerator.cs [hunk 2]
```diff
@@ -407,13 +389,16 @@ private string BuildPlugin(RoslynPluginDefinition definition, string baseTempDir
             string fullJarPath = Path.Combine(outputDirectory,
                 definition.Manifest.Key + "-plugin-" + definition.Manifest.Version + ".jar");
 
-            string repoKey = RepositoryKeyUtilities.GetValidKey(definition.PackageId + "." + definition.Language);
+            string repositoryId = RepositoryKeyUtilities.GetValidKey(definition.PackageId + "." + definition.Language);
+
+            string repoKey = RepositoryKeyPrefix + repositoryId;
 
-            RulesPluginBuilder builder = new RulesPluginBuilder(logger);
+            RoslynPluginJarBuilder builder = new RoslynPluginJarBuilder(logger);
```
Comment (duncanpMS) [inline RoslynPluginGenerator/AnalyzerPluginGenerator.cs]: I created a new RoslynPluginJarBuilder class with a very similar API to the old RulesPluginBuilder to minimise the change in the caller. I'll probably rename and move the new class out of the Roslyn-specific assembly, as it doesn't contain any Roslyn-specific code.

File: RoslynPluginGenerator/AnalyzerPluginGenerator.cs [hunk 3]
```diff
@@ -430,33 +415,16 @@ private string BuildPlugin(RoslynPluginDefinition definition, string baseTempDir
...
-                builder.AddExtension(extension);
-            }
+            builder.SetPluginProperty(repositoryId + ".nuget.packageId", definition.PackageId);
+            builder.SetPluginProperty(repositoryId + ".nuget.packageVersion", definition.PackageVersion);
+
+            builder.SetPluginProperty(repositoryId + ".analyzerId", definition.PackageId);
+            builder.SetPluginProperty(repositoryId + ".ruleNamespace", definition.PackageId);
+            builder.SetPluginProperty(repositoryId + ".staticResourceName", definition.StaticResourceName);
+            builder.SetPluginProperty(repositoryId + ".pluginKey", definition.Manifest.Key);
+            builder.SetPluginProperty(repositoryId + ".pluginVersion", definition.Manifest.Version);
```
Comment (duncanpMS) [inline RoslynPluginGenerator/AnalyzerPluginGenerator.cs]: The "SetPluginProperty" replaces the old code which injected the property values into source code that was then compiled. The properties are now written to the new configuration.xml file which is then embedded in the jar.

File: RoslynPluginGenerator/ArchiveUpdater.cs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+
+namespace SonarQube.Plugins.Roslyn
+{
+    /// <summary>
+    /// Updates an existing archive (zip, jar) by inserting additional files
+    /// </summary>
+    public class ArchiveUpdater
```
Comment (duncanpMS) [inline RoslynPluginGenerator/ArchiveUpdater.cs]: New helper class that handles inserting new files into an existing jar file.

File: RoslynPluginGenerator/ArchiveUpdater.cs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+        public ArchiveUpdater AddFile(string sourceFilePath, string relativeTargetFilePath)
+        {
+            this.fileMap[relativeTargetFilePath] = sourceFilePath;
+            return this;
+        }
+
+        public void UpdateArchive()
+        {
+            this.logger.LogDebug(UIResources.ZIP_UpdatingArchive, this.inputArchiveFilePath);
```
Comment (duncanpMS) [inline RoslynPluginGenerator/ArchiveUpdater.cs]: TODO: need to add some validation that the required properties have been set.

File: RoslynPluginGenerator/ArchiveUpdater.cs [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+                this.logger.LogDebug(UIResources.ZIP_DeletingExistingArchive);
+                File.Delete(this.outputArchiveFilePath);
+            }
+
+            ZipUsingShell(unpackedDir, this.outputArchiveFilePath);
+
+            this.logger.LogDebug(UIResources.ZIP_NewArchiveCreated, this.outputArchiveFilePath);
+        }
+        
+        private static void ZipUsingShell(string sourceDir, string targetZipFilePath)
```
Comment (duncanpMS) [inline RoslynPluginGenerator/ArchiveUpdater.cs]: This method does effectively the same as the PowerShell script in the MSBuild scanner packaging project i.e. uses a COM shell class to re-zip the folder (using the .Net ZipFile class creates an invalid jar).
I'll add a pointer to the Scanner PowerShell script and add a comment to this method.

File: RoslynPluginGenerator/DataModel/PluginProperties.cs [hunk 7]
```diff
@@ -16,21 +17,21 @@ namespace SonarQube.Plugins.Roslyn
     /// Describes a collection of Maven POM properties
     /// </summary>
     /// <remarks>This class is XML-serializable</remarks>
-    public class PluginProperties : StringDictionary, IXmlSerializable
+    public class PluginProperties : Dictionary<string, string>, IXmlSerializable
```
Comment (duncanpMS) [inline RoslynPluginGenerator/DataModel/PluginProperties.cs]: It turns out that StringDictionary forces all of the keys to lower-case, which isn't what we want, so I switched to deriving from Dictionary<> instead. I've updated the unit tests to check the case of the keys is preserved.

File: RoslynPluginGenerator/RoslynPluginDefinition.cs [hunk 8]
```diff
@@ -10,7 +10,7 @@ namespace SonarQube.Plugins.Roslyn
     /// <summary>
     /// Data class: contains the information required to create a Roslyn analyzer plugin
     /// </summary>
-    internal class RoslynPluginDefinition
+    public class RoslynPluginDefinition
```
Comment (duncanpMS) [inline RoslynPluginGenerator/RoslynPluginDefinition.cs]: Not sure this change is actually required: I'll check.

File: RoslynPluginGenerator/RoslynPluginJarBuilder.cs [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+namespace SonarQube.Plugins.Roslyn
+{
+    /// <summary>
+    /// Creates a SonarQube rules plugin
+    /// </summary>
+    public class RoslynPluginJarBuilder
```
Comment (duncanpMS) [inline RoslynPluginGenerator/RoslynPluginJarBuilder.cs]: The structure and public API of this class is very similar to the RulesPluginBuilder class it replaces.
The main difference is in how it builds the jar.

File: RoslynPluginGenerator/RoslynPluginJarBuilder.cs [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+
+
+        // Locations in the jar where various file should be embedded
+        private const string RelativeManifestResourcePath = "META-INF\\MANIFEST.MF";
+        private const string RelativeConfigurationResourcePath = "org\\sonar\\plugins\\roslynsdk\\configuration.xml";
+        private const string RelativeRulesXmlResourcePath = "org\\sonar\\plugins\\roslynsdk\\rules.xml";
+        private const string RelativeSqaleXmlResourcePath = "org\\sonar\\plugins\\roslynsdk\\sqale.xml";
+
+        private readonly ILogger logger;
+        private readonly ISet<string> sourceFiles;
```
Comment (duncanpMS) [inline RoslynPluginGenerator/RoslynPluginJarBuilder.cs]: No longer required. I'll remove it.

File: RoslynPluginGenerator/RoslynPluginJarBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+        /// </summary>
+        private void EnsureCoreManifestPropertiesExist()
+        {
+            // This property must appear first in the manifest.
+            // See http://docs.oracle.com/javase/6/docs/technotes/guides/jar/jar.html#JAR%20Manifest
+            this.manifestProperties.Add("Manifest-Version", "1.0");
+
+            this.manifestProperties.Add("Sonar-Version", "4.5.2");
+            this.manifestProperties.Add("Plugin-Dependencies", "META-INF/lib/sslr-squid-bridge-2.6.jar");
+            this.manifestProperties.Add("Plugin-SourcesUrl", "https://github.com/SonarSource-VisualStudio/sonarqube-roslyn-sdk-template-plugin");
```
Comment (duncanpMS) [inline RoslynPluginGenerator/RoslynPluginJarBuilder.cs]: These properties will need to change when the template jar changes.

File: RoslynPluginGenerator/SonarQube.Plugins.Roslyn.PluginGenerator.csproj [hunk 12]
```diff
@@ -178,6 +180,20 @@
...
+  <ItemGroup>
+    <COMReference Include="Shell32">
+      <Guid>{50A7E9B0-70EF-11D1-B75A-00A0C90564FE}</Guid>
+      <VersionMajor>1</VersionMajor>
+      <VersionMinor>0</VersionMinor>
+      <Lcid>0</Lcid>
+      <WrapperTool>tlbimp</WrapperTool>
+      <Isolated>False</Isolated>
+      <EmbedInteropTypes>True</EmbedInteropTypes>
+    </COMReference>
```
Comment (duncanpMS) [inline RoslynPluginGenerator/SonarQube.Plugins.Roslyn.PluginGenerator.csproj]: The Shell.Folder class is a COM object. This ref is to the type library that defines it. *EmbedInteropTypes* is true, which means the required COM interface definitions will be embedded into our assembly which it is built.

File: Tests/IntegrationTests/Roslyn/RoslynGenTests.cs [hunk 13]
```diff
@@ -29,7 +27,6 @@ public void RoslynPlugin_GenerateForValidAnalyzer_Succeeds()
             // Arrange
             TestLogger logger = new TestLogger();
             string outputDir = TestUtils.CreateTestDirectory(this.TestContext, ".out");
-            PluginInspector inspector = CreatePluginInspector(logger);
```
Comment (duncanpMS) [inline Tests/IntegrationTests/Roslyn/RoslynGenTests.cs]: Fixed up this test to not use the JDK.
Previously, we built and executed the Java PluginInspector class, which examined the new .jar, created the extensions, and dumped data about them to an XML now.
Now, we just check that the jar contains the expected XML files, and that those files have the expected content.

File: Tests/IntegrationTests/Roslyn/RoslynGenTests.cs [hunk 14]
```diff
@@ -151,110 +145,69 @@ private static string AssertPluginJarExists(string rootDir)
...
         }
 
-        private static void AssertRepositoryIsValid(JarInfo.Repository repository)
+        private static void AssertExpectedRulesExist(DiagnosticAnalyzer analyzer, string actualRuleFilePath)
         {
-            Assert.IsNotNull(repository, "Repository should not be null");
-
-            AssertPropertyHasValue(repository.Key, "Repository Key");
-            AssertPropertyHasValue(repository.Name, "Repository Name");
+            //XDocument rulesXml = XDocument.Load(actualRuleFilePath);
```
Comment (duncanpMS) [inline Tests/IntegrationTests/Roslyn/RoslynGenTests.cs]: TODO: once I've picked up @dbolkensteyn 's changes to the metadata classes, it should be possible to load the rules file into a strongly-typed RulesModel.

File: Tests/RoslynPluginGeneratorTests/ArchiveUpdaterTests.cs [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+{
+    [TestClass]
+    public class ArchiveUpdaterTests
+    {
+        public TestContext TestContext { get; set; }
+
+        #region Tests
+
+        [TestMethod]
+        public void ArchiveUpdater_SimpleUpdateJar_Succeeds()
```
Comment (duncanpMS) [inline Tests/RoslynPluginGeneratorTests/ArchiveUpdaterTests.cs]: Simple test for the zip updater: checks new files can be created in the root directory and sub-directories.

File: RoslynPluginGenerator/ArchiveUpdater.cs [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+                System.Threading.Thread.Sleep(copyPauseInMilliseconds);
+            }
+
+            // Rename the file to the expected name
+            File.Move(zipFilePath, targetZipFilePath);
+        }
+
+        private static void CreateEmptyZipFile(string filePath)
+        {
+            byte[] emptyZipHeader = new byte[] { 80, 75, 5, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
```
Comment (dbolkensteyn) [inline RoslynPluginGenerator/ArchiveUpdater.cs]: looks like we're not very far from re-implemeting our own Zip library :)

File: RoslynPluginGenerator/ArchiveUpdater.cs [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+
+            Type shellAppType = Type.GetTypeFromProgID("Shell.Application");
+            Object app = Activator.CreateInstance(shellAppType);
+
+            Shell32.Folder folder = shellAppType.InvokeMember("NameSpace", System.Reflection.BindingFlags.InvokeMethod, null, app, new object[] { zipFilePath }) as Shell32.Folder;
+
+            foreach(string dir in Directory.GetDirectories(sourceDir))
+            {
+                folder.CopyHere(dir, copyFlags);
+                System.Threading.Thread.Sleep(copyPauseInMilliseconds);
```
Comment (dbolkensteyn) [inline RoslynPluginGenerator/ArchiveUpdater.cs]: this doesn't seem to be ideal

File: RoslynPluginGenerator/RoslynPluginJarBuilder.cs [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+        }
+
+        /// <summary>
+        /// Sets a property that will appear in the manifest file
+        /// </summary>
+        public RoslynPluginJarBuilder SetManifestProperty(string name, string value)
+        {
+            if (string.IsNullOrWhiteSpace(name))
+            {
+                throw new ArgumentNullException("name");
```
Comment (dbolkensteyn) [inline RoslynPluginGenerator/RoslynPluginJarBuilder.cs]: `nameof()`?

File: RoslynPluginGenerator/RoslynPluginJarBuilder.cs [hunk 19]
```diff
@@ -0,0 +1,394 @@
...
+            this.fileToRelativePathMap = new Dictionary<string, string>();
+
+            this.EnsureCoreManifestPropertiesExist();
+        }
+
+        public RoslynPluginJarBuilder SetJarFilePath(string filePath)
+        {
+            if (string.IsNullOrWhiteSpace(filePath))
+            {
+                throw new ArgumentNullException("name");
```
Comment (dbolkensteyn) [inline RoslynPluginGenerator/RoslynPluginJarBuilder.cs]: yup, `nameof()` is definitely better :)

File: Tests/IntegrationTests/Roslyn/RoslynGenTests.cs (hunk 14 above)
Comment (dbolkensteyn) [inline Tests/IntegrationTests/Roslyn/RoslynGenTests.cs]: I've only changed slightly the `SqaleModel`, the rules one was left unchanged: https://github.com/SonarSource-VisualStudio/sonarqube-roslyn-sdk/blob/master/PluginGenerator/DataModel/Rules.cs#L17

**Comment by dbolkensteyn** [comment]
LGTM

I've manually checked the generated .jar for SonarAnalyzer, and tested the end-to-end on Wintellect.

File: RoslynPluginGenerator/ArchiveUpdater.cs (hunk 17 above)
Comment (duncanpMS) [inline RoslynPluginGenerator/ArchiveUpdater.cs]: It's not, but the copy is done on a different thread and we have no way of knowing that it has finished.

File: RoslynPluginGenerator/RoslynPluginJarBuilder.cs (hunk 19 above)
Comment (duncanpMS) [inline RoslynPluginGenerator/RoslynPluginJarBuilder.cs]: That would be a handy analysis rule/code fix (if there isn't one already)

Comment (dbolkensteyn) [inline RoslynPluginGenerator/RoslynPluginJarBuilder.cs]: :+1: @tvsonar ?

Comment (tvsonar) [inline RoslynPluginGenerator/RoslynPluginJarBuilder.cs]: Created a ticket for it: https://jira.sonarsource.com/browse/SLVS-824

File: RoslynPluginGenerator/ArchiveUpdater.cs (hunk 16 above)
Comment (duncanpMS) [inline RoslynPluginGenerator/ArchiveUpdater.cs]: It's definitely more complicated than we'd want :-(

**Pull request merged**